# 4.3 Geocoding

<br></br>
<font size="3">What is geocoding?</font>

<br></br>

<font size="3">Geocoding is to converti addresses into coordinates (or vice versa), which is a really common GIS task. For example, we might want to know the longitude and latitude from a business address. Fortunately, there are convinient libraries in python that make the geocoding task really easy.</font>



<br></br>
<font size="3">One of the libraries that can do the geocoding for us is `geopy` (https://geopy.readthedocs.io/en/stable/#module-geopy.geocoders) that makes it easy to locate the coordinates of addresses, cities, countries, and landmarks across the globe using third-party geocoders and other data sources. You can think about `geopy` as a "wrapper"; however, what's inside is geocoder services, and many of them are not FREE! (also can be very expensive if you have large queries). Some of the providers are: </font>

* ArcGIS online geocoding service (https://geocode.arcgis.com).
* Google Geocoding API (https://developers.google.com/maps/documentation/geocoding/start).
* OpenStreetMap (Nominatim) (https://nominatim.openstreetmap.org).




<br></br>
<font size="3">Great thing about `geopy` is that it is integrated in `geopandas`, which makes it very friendly (native) to use. Now let's give it a try.</font>





In [10]:
my_address = "1037 South Farmer Ave, Tempe, Arizona"

<br></br>
<font size="3">Above is my apartment address, is someone living close to me? </font>
<br></br>
<font size="3">And next, let's try to convert this address to coordinates.The command is: </font>


<br></br>
<font size="3">`geopandas.tools.geocode.eocode(my_address, provider='nominatim')` </font>
    
<font size="3">where we use the OpenStreetMap's Nominatim's geocoding service.


In [11]:
from geopandas.tools import geocode
# Geocode addresses with Nominatim backend
coords = geocode(my_address, provider='nominatim')

In [12]:
coords

address  \
0  1037, South Farmer Avenue, Tempe Junction, Tem...   

                                     geometry  
0  POINT (-111.944135926829 33.4176156097561)



<br></br>
<font size="3">What we get is a GeoDataFrame with only one row that contains my address and the coordinates, which seems to be quite accurate. (you can validate it on Google Maps)</font>

<br></br>
<font size="3">`POINT (-111.944135926829 33.4176156097561)`</font>



<br></br>
<font size="3">What if we have multiple addresses? We just need to give the `geocode` function a list of addresses. </font>
<br></br>
<font size="3">For example, if we have two locations: My home and Wells Fargo Arena. </font>



In [44]:
my_address_list = ["1037 South Farmer Ave, Tempe, Arizona", 
                   "Wells Fargo Arena, Tempe, Arizona"]

coords_list = geocode(my_address_list, provider='nominatim')

In [45]:
coords_list

address  \
0  1037, South Farmer Avenue, Tempe Junction, Tem...   
1  Wells Fargo Arena, South Stadium Drive, Tempe ...   

                                     geometry  
0  POINT (-111.944135926829 33.4176156097561)  
1       POINT (-111.931008934445 33.42451785)

<font size="3">That being said, if we have a Excel table of addresses and needed be geocoded, we can do this in a batch. </font>

<font size="3">Let's try to plot those two locations on a map using the coordinates we obtained from the geocoder (I will talk about how to make a map like below  in more details in the Module 6). </font>

In [48]:
import folium
%matplotlib inline
asu = folium.Map(
    location=[33.4200, -111.9350],
    tiles='Stamen Toner',
    zoom_start=15
)
folium.Marker([coords_list.geometry[1].y, coords_list.geometry[1].x], popup='<i>Arena</i>').add_to(asu)
folium.Marker([coords_list.geometry[0].y, coords_list.geometry[0].x], popup='<i>Home</i>').add_to(asu)
asu